pkg/apiserver/handlers/webhook/mutation_labels.go

```diff
@@ -0,0 +1,175 @@
+		if contentType != "application/json" {
+			klog.Errorf("Invalid Content-Type=%s, expected application/json", contentType)
+			http.Error(w, "invalid Content-Type, expected `application/json`", http.StatusUnsupportedMediaType)
+			return
+		}
+		var admissionResponse *admv1.AdmissionResponse
+		ar := admv1.AdmissionReview{}
+		ar.TypeMeta.Kind = "AdmissionReview"
+		ar.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if err := json.Unmarshal(reqBody, &ar); err != nil {
+			klog.Errorf("Label mutation received incorrect body")
+			admissionResponse = getAdmissionResponseForErr(err)
+		} else {
+			admissionResponse = mutateResourceLabels(&ar)
+		}
+		aReview := admv1.AdmissionReview{}
+		aReview.TypeMeta.Kind = "AdmissionReview"
+		aReview.TypeMeta.APIVersion = "admission.k8s.io/v1"
+		if admissionResponse != nil {
+			aReview.Response = admissionResponse
+			if ar.Request != nil {
+				aReview.Response.UID = ar.Request.UID
+			}
+		}
+		resp, err := json.Marshal(aReview)
+		if err != nil {
+			klog.Errorf("Unable to encode response during mutation: %v", err)
+			http.Error(w, fmt.Sprintf("could not encode response: %v", err), http.StatusInternalServerError)
+		}
+		klog.V(2).Infof("Writing mutation response to MutationAdmissionHook")
+		if _, err := w.Write(resp); err != nil {
+			klog.Errorf("Unable to write response during mutation: %v", err)
+			http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+		}
+	}
+}
+
+// mutateResourceLabels mutates the resource labels and inserts Antrea required labels.
+func mutateResourceLabels(ar *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var patch []byte
+	var allowed bool
+	var msg string
+	// At the moment we only mutate Namespace labels.
+	var nsObj v1.Namespace
+	patchType := admv1.PatchTypeJSONPatch
+
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	klog.V(2).Info("Mutating resource labels")
+	if curRaw != nil {
+		if err := json.Unmarshal(curRaw, &nsObj); err != nil {
+			klog.Errorf("Error de-serializing current resource")
+			return getAdmissionResponseForErr(err)
+		}
+	}
+	msg, allowed, patch = mutateLabels(op, nsObj.Labels, nsObj.Name)
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed:   allowed,
+		Result:    result,
+		PatchType: &patchType,
+		Patch:     patch,
+	}
+}
+
+// mutateLabels mutates the resource's labels and forcefully inserts the resource's name as a well
+// known label to ensure that the label is never modified or removed by CREATE and UPDATE events.
+func mutateLabels(op admv1.Operation, l map[string]string, name string) (string, bool, []byte) {
+	var patch []byte
+	switch op {
+	case admv1.Create, admv1.Update:
+		if l == nil {
+			l = map[string]string{}
+		}
+		// Forcefully stomp the resource's metadata.name value as a label.
+		l[LabelMetadataName] = name
+		genPatch, err := createLabelsReplacePatch(l)
```

Comment:
I think it was because the `patch` var is declared on a outside scope, and `err` is a new variable, so changing `genPatch` to `patch` in this line will not assign to the outer-scoped `patch`. Refactored it anyways.